### DISCUSSION TITLE
Different behavior when  cache_dir defined in env

### DISCUSSION THREAD

**Discussion by Delsian** [ccache/ccache#1544] [Questions & answers]
I've observed a strange issue with my build. It works fine with the default ccache config but fails if I set CCACHE_DIR environment, even if it points to the same folder:

```
$ ccache -p
(default) absolute_paths_in_stderr = false
(default) base_dir = 
(default) cache_dir = /home/eug/.cache/ccache
...
[Here all works ok]
$ export CCACHE_DIR=/home/eug/.cache/ccache
$ ccache -p
(default) absolute_paths_in_stderr = false
(default) base_dir = 
(environment) cache_dir = /home/eug/.cache/ccache
[and build fails here]
```

Looks like some config files reads as empty.
How it is possible if ccache has the same configuration? Nothing changed except environment, effect 100% reproducible.

What I can check/change in this situation?

Ubuntu 24.04  ccache version 4.9.1

**Comment by jrosdahl** [comment]
The default location of the configuration file is `~/.config/ccache/ccache.conf`. However, if you set `CCACHE_DIR` then the configuration file is read from `$CCACHE_DIR/ccache.conf`. This is for backward compatibility reasons. More details here: https://ccache.dev/manual/4.10.2.html#_location_of_the_configuration_file

You can also see this difference if you run `ccache -sv`:

```console
$ ccache -sv | grep ^Config   
Config file:        /home/joel/.config/ccache/ccache.conf
$ CCACHE_DIR=/home/joel/.cache/ccache ccache -sv | grep ^Config
Config file:        /home/joel/.cache/ccache/ccache.conf
```

**Reply by Delsian** [reply]
And how can it affect, if I have no ccache.conf file in both locations (and in my case - the same area)?
```
eug@eug-HP ~ $ ls -la ~/.config/ccache/ccache.conf
ls: cannot access '/home/eug/.config/ccache/ccache.conf': No such file or directory
eug@eug-HP ~ $ 
```

**Reply by jrosdahl** [reply]
OK, if you don't have any configuration file then the behavior shouldn't be different if you set `CCACHE_DIR`.

But then, why did you say "Looks like some config files reads as empty"? What did you observe? And how did the build fail?

**Reply by Delsian** [reply]
During the build I have a cmake script run over the source files collecting objects to link. When I set CCACHE_DIR, I see some files are not present in the generated list. In the log I saw that files was opened by script. Without CCACHE_DIR everything works as expected. The only change between runs is an environment variable.

**Reply by jrosdahl** [reply]
I suggest setting `CCACHE_LOGFILE` to a log file. Then compare logs from with and without `CCACHE_DIR` set to see what is different.
